### PR TITLE
Introduce WEEK interval

### PR DESCRIPTION
**Pull request by erickmp07**
Related to #53282 

This PR implements the WEEK interval to Histogram.

### CODE REVIEW

**Review by bpintea — CHANGES_REQUESTED** [review]
Thanks for the contribution and sorry for the late feedback.
The implementation seems good to me, but could you please add some more [tests](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/sql/qa/server/src/main/resources/agg.csv-spec)  and complete the [docs](https://github.com/elastic/elasticsearch/blob/master/docs/reference/sql/functions/grouping.asciidoc)?

**Comment by erickmp07** [comment]
@bpintea I think I made all the requested changes.
If something is missing, please let me know.

**Comment by bpintea** [comment]
@elasticmachine ok to test

File: docs/reference/sql/functions/grouping.asciidoc [hunk 1]
```diff
@@ -87,8 +87,8 @@ actually used will be `INTERVAL '2' DAY`. If the interval specified is less than
 
 [IMPORTANT]
 All intervals specified for a date/time HISTOGRAM will use a <<search-aggregations-bucket-datehistogram-aggregation,fixed interval>>
-in their `date_histogram` aggregation definition, with the notable exceptions of `INTERVAL '1' YEAR`, `INTERVAL '1' MONTH` and `INTERVAL '1' DAY`  where a calendar interval is used.
-The choice for a calendar interval was made for having a more intuitive result for YEAR, MONTH and DAY groupings. In the case of YEAR, for example, the calendar intervals consider a one year
+in their `date_histogram` aggregation definition, with the notable exceptions of `INTERVAL '1' YEAR`, `INTERVAL '1' MONTH`, `INTERVAL '1' DAY` and `INTERVAL '1' WEEK`  where a calendar interval is used.
```
Comment (bpintea) [inline docs/reference/sql/functions/grouping.asciidoc:90]: nit: could you please place the week reference in sequence, as you did below?

**Review by bpintea — COMMENTED** [review]
There seem to still be some query translation test failures, you might want to run the tests and address those please.

**Review by matriv — CHANGES_REQUESTED** [review]
This PR only adds the WEEK Interval functionality for histograms, it doesn't add it as a new interval type. We cannot have queries like `SELECT * FROM test WHERE my_date >= now() - INTERVAL 1 WEEK.

Furthermore, what will happen if you have a histogram with `INTERVAL 7 DAYS`? As far as I can understand this will be translated to a calendar interval of `1w` which is not what we want, as 7 DAYS should be an absolute interval.

**Comment by erickmp07** [comment]
@matriv The issue description mentions that week interval implementation could be useful and could help with scenarios similar to #40162 and #52713

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 2]
```diff
@@ -1398,6 +1398,30 @@ public void testGroupByMoreDaysHistogram_WithFunction_QueryTranslator() {
...
+    public void testGroupByMoreWeekHistogramQueryTranslator() {
+        PhysicalPlan p = optimizeAndPlan("SELECT HISTOGRAM(date, INTERVAL 3 WEEK) AS h FROM test GROUP BY h");
+        assertEquals(EsQueryExec.class, p.getClass());
+        EsQueryExec eqe = (EsQueryExec) p;
+        assertEquals(1, eqe.output().size());
+        assertEquals("h", eqe.output().get(0).qualifiedName());
+        assertEquals(DATETIME, eqe.output().get(0).dataType());
+        assertThat(eqe.queryContainer().aggs().asAggBuilder().toString().replaceAll("\\s+", ""),
+            endsWith("\"date_histogram\":{\"field\":\"date\",\"missing_bucket\":true,\"value_type\":\"date\",\"order\":\"asc\","
+                    + "\"calendar_interval\":\"1814400000ms\",\"time_zone\":\"Z\"}}}]}}}"));
```
Comment (matriv) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:1422]: This is not correct, it should be `\"calendar_interval\":\"3w\"`.

Comment (erickmp07) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: Fixed. Thanks.
I tried to run the tests here, but I've got some gradle errors.
Command:
`./gradlew test --tests org.elasticsearch.xpack.sql.planner.QueryTranslatorTests`

Result:
`Task :test:logger-usage:test FAILED`
`Task :libs:elasticsearch-core:test FAILED`
`Task :build-tools:test FAILED`

Do you have some suggestion to fix it?

**Comment by matriv** [comment]
@erickmp07 

First of all, this implementation is an incomplete approach. As I've already mentioned, It adds the support for WEEK interval only for histograms and not as a complete type which you can use for other types of calculations and filterings. Moreover, this incomplete approach is also wrong because of this: https://github.com/elastic/elasticsearch/pull/59335/files#diff-4d4ca1fe9d2bf59cd1e9e61889fa691cR349 The unit tests are failing because with this check you cannot achieve multiple week intervals like you have in the test for 3 weeks, and also as I've mentioned already, you cannot distinguish between INTERVAL 7 DAYS and INTERVAL 1 WEEK which is a different histogram.

I'd like to thank you for your time and effort on this one, but I'm afraid I have to proceed with rejecting and closing this PR as it's currently not in a good state to continue the work on it. 

PS. Please also keep in mind that we've just marked the related issue as under discussion, since we need to spend some more time thinking and discussing if we should go ahead with this approach, as the WEEK interval is not supported in SQL standard.

Thanks again for your interest and effort so far.

**Pull request closed without merging**